### PR TITLE
Add supports for aquifers in parallel

### PR DESCRIPTION
**Pull request by akva2**
Downstream of https://github.com/OPM/opm-common/pull/2170

### CODE REVIEW

File: opm/simulators/aquifers/AquiferFetkovich.hpp
Comment (totto82) [inline opm/simulators/aquifers/AquiferFetkovich.hpp:68]: Is it more readable if this->W_flux_ is communicated here and not in aquiferPressure()

**Review by totto82 — APPROVED** [review]

**Comment by totto82** [comment]
This can be merged if the test are green.

File: ebos/eclbasevanguard.hh
Comment (totto82) [inline ebos/eclbasevanguard.hh:807]: why do you need to be a map?

Comment (akva2) [reply ebos/eclbasevanguard.hh:807]: To avoid a full cart size vector.. ie only store ids for local cells.

**Comment by blattms** [comment]
benchmark opm-common=2170 please

**Comment by akva2** [comment]
there is still something fishy with ct aquifers, so please hold merge.

File: ebos/eclbasevanguard.hh
Comment (blattms) [reply ebos/eclbasevanguard.hh:807]: Somebody should check whether using a map somewhere really saves anything. Let us do some back of the envelope calculation:

For each map entry you need to store key and value (for int that should be 8 bytes) + plus the overhead of red-black tree node (for gcc that is another 3 pointers (parent, left, right) + an enum, hence more than 12 bytes per entry).

A vector only needs 4 bytes per entry. Unless we usually have  less than 20% of active cells, std::map seems an odd choice even if performance does not matter at all. We probably can assume that if we have more than 4 processes computing, but below this threshold we are very generous with our memory.

IMHO at some time this needs to done properly using a sorted vector of std::pairs or an unordered_map. Personally, I would prefer the former (but I might be a bit lonely with this opinion), because once the map is set up no further entries will be added.

Comment (akva2) [reply ebos/eclbasevanguard.hh:807]: it's not just about active cells, it's about cells on this process vs global cartesian cells. as in, the vector is of full cartesian size (of the underlying grid). but anyways, i'll just drop this commit.

File: opm/simulators/aquifers/AquiferFetkovich.hpp
Comment (akva2) [reply opm/simulators/aquifers/AquiferFetkovich.hpp:68]: W_flux_ is an Evaluation so I cannot easily collect the global values. Thus I see no way to move the reduction.

**Comment by akva2** [comment]
jenkins build this please

File: opm/simulators/aquifers/AquiferInterface.hpp
Comment (atgeirr) [inline opm/simulators/aquifers/AquiferInterface.hpp:307]: While here anyway, we should make this a logic_error, as it should only arise through programmer error. Also, the error message is misleading, I suggest "Internal error in initialization of aquifer." or something like that.

File: opm/simulators/aquifers/AquiferCarterTracy.hpp
Comment (atgeirr) [inline opm/simulators/aquifers/AquiferCarterTracy.hpp:93]: Please add trailing underscore in line with other class variables here. Also, the `this->` convention is used for other members, so best to apply it to fluxValue as well.

File: opm/simulators/aquifers/AquiferInterface.hpp
Comment (atgeirr) [inline opm/simulators/aquifers/AquiferInterface.hpp:307]: This should have been a logic_error all along i believe, and the error message is misleading. Maybe just "Internal error in aquifer setup." or similar.

File: ebos/eclwriter.hh
Comment (atgeirr) [inline ebos/eclwriter.hh:417]: I do not understand why you do not add the aquifer data here?

**Review by atgeirr — COMMENTED** [review]
Only minor comments from me. I have read it quite carefully, except for the packunpack parts which I have only scanned superficially.

**Comment by alfbr** [comment]
This has been tested and confirmed correct. Please consider merging as there are other PR's in flight.

**Comment by alfbr** [comment]
jenkins build this please

**Comment by totto82** [comment]
> there is still something fishy with ct aquifers, so please hold merge.

Is this fixed now?

**Comment by akva2** [comment]
Yes

**Comment by akva2** [comment]
That being said there are two unadressed review comments i want to take care of before merge, give me an hour

File: opm/simulators/aquifers/AquiferCarterTracy.hpp
Comment (akva2) [reply opm/simulators/aquifers/AquiferCarterTracy.hpp]: done.

File: ebos/eclwriter.hh
Comment (akva2) [reply ebos/eclwriter.hh:417]: The aquifer data is only used in evalSummary, it is not used in writeOutput. so there is no need to communicate it here.

File: opm/simulators/aquifers/AquiferInterface.hpp
Comment (akva2) [reply opm/simulators/aquifers/AquiferInterface.hpp]: done.

**Comment by akva2** [comment]
jenkins build this please

File: ebos/eclwriter.hh
Comment (atgeirr) [reply ebos/eclwriter.hh:417]: Thanks, I assume that is because we only output aquifer data in summary-type output?

Comment (bska) [reply ebos/eclwriter.hh:417]: For now, yes. Adding aquifer restart output is on my list of things to do in the new year.

**Comment by akva2** [comment]
green, approved, comments adressed.

File: ebos/collecttoiorank.hh
Comment (GitPaean) [inline ebos/collecttoiorank.hh:766]: This will makes the parallel restarting more difficult.  Because with this, each process only have its own portion of `volume` (theoretically), when restarting, you only have the one for the whole aquifer, then it is not easy to recover the individual potions anymore. 

I suggest we handle the parallel summation within the simulation code (aquifer model code), then all the processes have the same `volume`. 

`fluxRate` is probably should be the same since we use `fluxRate` to update `volume`.

Comment (blattms) [reply ebos/collecttoiorank.hh:766]: I am not sure you are aware that this not connected directly to restart.
Unpackinging will be done on the root process during file output only. It will unpack one result for each process. The result is only written to disk. Hence only the root process needs to have the same result as a serial run (I guess that is assured by taking the max or the sum).

Comment (GitPaean) [reply ebos/collecttoiorank.hh:766]: > I am not sure you are aware that this not connected directly to restart.

It is not directly related to restart. But when considering restarting, how you can manage to get some meaningful `volume` for each process to begin with, because `volume` is an accumulative term.  Although at the end, it is the root process result that matters for output, but imagining you have some meaningful information on the root process, and other processes only have some half meaningful information after restarting. I do not think that is pretty.

Comment (bska) [reply ebos/collecttoiorank.hh:766]: For what it's worth, the restart file will have certain pieces of information pertaining to each aquifer-reservoir connection in addition to values aggregated to the aquifer as a whole. I haven't fully characterised everything yet. It **may** happen that we need to maintain additional dynamic quantities at the connection level.

**Pull request merged**
